FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_multitask/DemoTasks/MultitaskMQTTExample.c

```diff
@@ -0,0 +1,1457 @@
+            break;
+
+        default:
+            /* Other operations don't need a context. */
+            break;
+    }
+
+    return xIsValid;
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvAddCommandToQueue( Command_t * pxCommand )
+{
+    xQueueSend( xCommandQueue, pxCommand, DEMO_TICKS_TO_WAIT );
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvCopyPublishToQueue( MQTTPublishInfo_t * pPublishInfo,
+                                   QueueHandle_t pResponseQueue )
+{
+    PublishElement_t xCopiedPublish;
+    MQTTPublishInfo_t * pxCopiedPublishInfo = NULL;
+
+    memset( ( void * ) &xCopiedPublish, 0x00, sizeof( xCopiedPublish ) );
+    pxCopiedPublishInfo = &( xCopiedPublish.publishInfo );
+    memcpy( &( xCopiedPublish.publishInfo ), pPublishInfo, sizeof( MQTTPublishInfo_t ) );
+
+    /* Since adding an MQTTPublishInfo_t to a queue will not copy its string buffers,
+     * we need to add buffers to a struct and copy the entire structure. */
+    memcpy( xCopiedPublish.pTopicName, pPublishInfo->pTopicName, pPublishInfo->topicNameLength );
+    memcpy( xCopiedPublish.pPayload, pPublishInfo->pPayload, pPublishInfo->payloadLength );
+    pxCopiedPublishInfo->pTopicName = ( const char * ) xCopiedPublish.pTopicName;
+    pxCopiedPublishInfo->pPayload = xCopiedPublish.pPayload;
+
+    /* Add to response queue. */
+    xQueueSendToBack( pResponseQueue, pxCopiedPublishInfo, DEMO_TICKS_TO_WAIT );
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t prvProcessCommand( Command_t * pxCommand )
+{
+    MQTTStatus_t xStatus = MQTTSuccess;
+    uint16_t usPacketId = MQTT_PACKET_ID_INVALID;
+    bool xAddAckToList = false;
+    MQTTPublishInfo_t * pxPublishInfo;
+    MQTTSubscribeInfo_t * pxSubscribeInfo;
+
+    switch( pxCommand->commandType )
+    {
+        case PROCESSLOOP:
+            LogInfo( ( "Running Process Loop." ) );
+            xStatus = MQTT_ProcessLoop( &globalMqttContext, MQTT_PROCESS_LOOP_TIMEOUT_MS );
+            break;
+
+        case PUBLISH:
+            assert( pxCommand->pContext != NULL );
+            pxPublishInfo = pxCommand->pContext->pPublishInfo;
+            assert( pxPublishInfo != NULL );
+
+            if( pxPublishInfo->qos != MQTTQoS0 )
+            {
+                usPacketId = MQTT_GetPacketId( &globalMqttContext );
+            }
+
+            LogInfo( ( "Publishing message to %.*s.", ( int ) pxPublishInfo->topicNameLength, pxPublishInfo->pTopicName ) );
+            xStatus = MQTT_Publish( &globalMqttContext, pxPublishInfo, usPacketId );
+            pxCommand->pContext->returnStatus = xStatus;
+
+            /* Add to pending ack list, or call callback if QoS 0. */
+            xAddAckToList = ( pxPublishInfo->qos != MQTTQoS0 ) && ( xStatus == MQTTSuccess );
+            break;
+
+        /* TODO: Option to subscribe/unsubscribe without sending a packet,
```

Comment:
When is this TODO planning to be addressed?

Reply:
I don't think it is anymore, so I will remove it